
## Multi-class classification problem using wine dataset                                             
This dataset is result of chemical analysis of wines grown in Italy using three diffferent cultivars.                  
This dataset is available in scikit-learn library.

The dataset comprises 13 features and target.

In [1]:
# Load dataset
from sklearn import datasets
wine = datasets.load_wine()

### Exploratory data analysis

In [4]:
# Explore data
print(wine.feature_names)

['alcohol', 'malic_acid', 'ash', 'alcalinity_of_ash', 'magnesium', 'total_phenols', 'flavanoids', 'nonflavanoid_phenols', 'proanthocyanins', 'color_intensity', 'hue', 'od280/od315_of_diluted_wines', 'proline']


In [5]:
print(wine.target_names)

['class_0' 'class_1' 'class_2']


In [6]:
print(wine.data[0:5]) # print top 5 records of feature set

[[1.423e+01 1.710e+00 2.430e+00 1.560e+01 1.270e+02 2.800e+00 3.060e+00
  2.800e-01 2.290e+00 5.640e+00 1.040e+00 3.920e+00 1.065e+03]
 [1.320e+01 1.780e+00 2.140e+00 1.120e+01 1.000e+02 2.650e+00 2.760e+00
  2.600e-01 1.280e+00 4.380e+00 1.050e+00 3.400e+00 1.050e+03]
 [1.316e+01 2.360e+00 2.670e+00 1.860e+01 1.010e+02 2.800e+00 3.240e+00
  3.000e-01 2.810e+00 5.680e+00 1.030e+00 3.170e+00 1.185e+03]
 [1.437e+01 1.950e+00 2.500e+00 1.680e+01 1.130e+02 3.850e+00 3.490e+00
  2.400e-01 2.180e+00 7.800e+00 8.600e-01 3.450e+00 1.480e+03]
 [1.324e+01 2.590e+00 2.870e+00 2.100e+01 1.180e+02 2.800e+00 2.690e+00
  3.900e-01 1.820e+00 4.320e+00 1.040e+00 2.930e+00 7.350e+02]]


In [7]:
print(wine.target)  # print records of target set ie. wine labels

[0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1
 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1
 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2
 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2]


In [8]:
print(wine.data.shape)  # print data (feature) shape

(178, 13)


In [9]:
print(wine.target.shape) # print target (label) shape

(178,)


#### Splitting data
To understand model performance, dividing the dataset into training set and test set is good practice.           
Here we use 70% data for training and 30% data for test purpose.                                     
We use random_state=5 to generate exactly same outputs each time.

In [10]:
# Split dataset into training set and test set
from sklearn.model_selection import train_test_split

# 70% training and 30% test
x_train, x_test, y_train, y_test = train_test_split(wine.data, wine.target, test_size=0.3, random_state=5)

### Generating Model
Building KNN classifier model                      
Import KNeighborsClassifier module from sklearn and create classifier object.                                                
Number of nearest neighbors used here is K = 7.

In [11]:
from sklearn.neighbors import KNeighborsClassifier

# create KNN classifier
knn_model = KNeighborsClassifier(n_neighbors = 7)  # Take K = 7

# Train the model using training set
knn_model.fit(x_train, y_train)

# predict for test
y_pred = knn_model.predict(x_test)

### Model Evaluation
Let's estimate how accurately model can predict the type of cultivars. For this purpose, import metrics module from sklearn.

In [12]:
from sklearn import metrics

# Check Accuracy of model
accuracy = metrics.accuracy_score(y_test, y_pred)
print("Accuracy : ", accuracy)

Accuracy :  0.7592592592592593


Accuracy of model is 75.92% for K=7 which is considered as good accuracy.                             
For further evaluation, we can also create model for different number of neighbors.  

### Regenerating model
Let's build KNN classifier model for K=5.

In [14]:
from sklearn.neighbors import KNeighborsClassifier

# create KNN classifier
knn_model = KNeighborsClassifier(n_neighbors = 5)

# Train the model using training set
knn_model.fit(x_train, y_train)

# predict for test
y_pred = knn_model.predict(x_test)

# Check Accuracy of model
accuracy = metrics.accuracy_score(y_test, y_pred)
print("Accuracy : ", accuracy)

Accuracy :  0.7037037037037037


Accuracy of model is 70.37% for K=5 which is also a good accuracy.